### ISSUE TITLE
CLI for ANN

### ISSUE THREAD

**Issue by arjunmenon** [mlpack/mlpack#1254]
Hey
Are ANN classes available from the CLI? From the docs it isn't apparent.
How can we use it otherwise?

**Comment by zoq** [comment]
You are right there doesn't exist an CLI for the ANN classes. You would have to use it from within your C++ code. I think it would be great to have an executable for the network code, however since there is no single architecture it's somewhat complex to provide all the necessary settings and at the same time make it easy to be used.

**Comment by arjunmenon** [comment]
I am not proficient with C++. I was hoping to add a Ruby wrapper to the CLI. So two things, 
1. Is there support planned for a Ruby binding
2. If I have to make an attempt, where do I start.

**Comment by zoq** [comment]
For more information about how to add bindings to other languages please take a look at: http://www.mlpack.org/docs/mlpack-git/doxygen/bindings.html, let us know if we should clarify anything.

**Comment by rcurtin** [comment]
Hi Arjun,

There isn't currently any planned support for Ruby.  If you are interested in writing a Ruby binding generator, it would be great and @zoq has given a link to some useful documentation.  But unfortunately proficiency with C++ is going to be necessary to write this binding generator, so you may want to also study C++ a little more closely before diving too deep.

**Comment by desai-aditya** [comment]
Hi rcurtin,

I was planning to work on this in small pieces. Could I first implement a CLI that achieves this -
https://github.com/mlpack/models/blob/master/Kaggle/DigitRecognizer/src/DigitRecognizer.cpp
and then later improve it?

**Comment by rcurtin** [comment]
We need to wait for @zoq's input also, but my opinion is that any CLI program needs the following functionality:

 * specify a neural network architecture arbitrarily (possibly by loading it from a YAML file or something?  We should probably discuss that part)
 * train on some given training data, with the user able to specify the optimizer and optimizer parameters
 * evaluate a trained model on some given testing data
 * print performance measures on training/test data
 * load an existing model or save a newly trained model

So writing something for a digit recognizer might be a good warmup exercise, but anything that we merge into mlpack should at least have the requirements above, in my opinion.

**Comment by akhandait** [comment]
@zoq, since we don't have any CLI for ANN classes, do you think it will be a good idea to start with a CLI for fully connected ANN with options for all other required parameters?
If yes, I would like to work on it.
@rcurtin I think it will be possible to incorporate all the other functionality mentioned above.
Maybe we could further develop it to support any neural network architecture.

**Comment by desai-aditya** [comment]
@rcurtin : 
What do you mean by 'loading it from a YAML file or something'?

**Comment by zoq** [comment]
I agree with @rcurtin, we should not merge anything that doesn't support:

> - specify a neural network architecture arbitrarily (possibly by loading it from a YAML file or something? We should probably discuss that part)
> - train on some given training data, with the user able to specify the optimizer and optimizer parameters
> - evaluate a trained model on some given testing data
> - print performance measures on training/test data
> - load an existing model or save a newly trained model

YAML for the config file sounds like a good idea to me, however parsing isn't as simple as e.g. csv and I don't like to include another dependency (perhaps boost spirit could be helpful).

> What do you mean by 'loading it from a YAML file or something'?

Instead of specifying the architecture e.g. layers via command line parameters we use a config file, YAML is just the format, we could also use JSON or XML.

> Also how about adding more activation functions?

If you have something in mind, we are open for suggestions.

> why is leakyReLU in the layers directory?

The LeakyReLU layer doesn't follow the same interface as e.g. the TanH since it holds an extra parameter ``alpha``.

**Comment by rcurtin** [comment]
> YAML for the config file sounds like a good idea to me, however parsing isn't as simple as e.g. csv and I don't like to include another dependency (perhaps boost spirit could be helpful).
>
>> What do you mean by 'loading it from a YAML file or something'?
>
> Instead of specifying the architecture e.g. layers via command line parameters we use a config file, YAML is just the format, we could also use JSON or XML.

Agreed, I want to avoid new dependencies also.  It's also possible we could come up with a very simple custom format that we can parse ourselves.  I do think boost::spirit could be used to parse YAML but it looks like implementing a YAML parser might be complex:

https://github.com/cierelabs/yaml_spirit

So maybe some plain text file is best?

```
linear 50 50
sigmoid
linear 50 50
sigmoid
linear 50 10
sigmoid
```

could be one way to do it for a 3-layer linear network with 50 hidden neurons in each hidden layer and 10 output neurons.  I don't have much of a preference; personally I think we could do just about anything here and as long as it is well documented and simple enough then nobody will have a problem with it.

**Comment by zoq** [comment]
Agreed, that's super simple and easy to read.

**Comment by akhandait** [comment]
I really like this idea. Also, I have worked with neural networks before and would love to write a CLI for it.
I think it will be really good and neat to input the network configuration as a txt file.
Can I go ahead and start implementing this, keeping all the 5 requirements in mind? :)

**Comment by akhandait** [comment]
@desai-aditya Do you want to work on this as well?
It also doesn't seem to be a small project, will you like it if we worked on it together?

**Comment by desai-aditya** [comment]
Yes we can surely collaborate @akhandait . How do you suggest we split the work?

**Comment by akhandait** [comment]
Okay, so I think we will need a day or two just to get really comfortable with the ANN method. The next step according to me will be to come with a concrete model for how we will implement it. If you have already dived into it and are comfortable maybe you could look at the CLI bindings we already have and come up with a basic structure for our implementation.
After that maybe we can discuss how to split the work.
Let's keep each other updated here on this issue. 
Let's do this!
@rcurtin @zoq If you feel like we could do this a better way, please tell.

**Comment by rcurtin** [comment]
This all sounds good to me.  Probably we will have to flesh out the file configuration format somewhat, but if we can keep it of the form

```
<layer type> <input size> <output size> <other parameters...>
```

I think that would be simplest to parse and work with.  Sometimes the input or output size may not be needed.  It'll also be really important to document exactly what the format is so that people can assemble simple networks just by looking at the documentation.

**Comment by akhandait** [comment]
Yes, 
@desai-aditya Documenting it well should be one of our top priorities.
Maybe we could also add some tutorials but that is thinking quite into the future.

**Comment by desai-aditya** [comment]
so help me if I forgot something -
first we'll just focus on linear models and then later add rnns
<batch size> : any int  (-b)
<train-test split ratio> : double between 0 and 1 (-r)
<optimizer> : type of optimizer: sgd , I dont know which others can fit need to read docs (-op)
<step size of sgd> : double  (-s)
<test file> : filename (-t)
<output file> :  filename (-o)
<train file> : filename (-i)
<print performance> :  (-v)

and the following in network.conf file
for each layer
<layer type> : linear
<input size> : any int but the first layer must match the dimensions of the dataset
<output size> : any int
<num epochs> : any int

is this fine? @zoq, @rcurtin ,@akhandait

**Comment by zoq** [comment]
Looks good to me, and I agree let's start with the linear models. About the network.conf file, ideally we interpret everything after the layer name as parameter, like for the linear layer this is input and output size, but for the convolution layer, there are more parameter we could set.

We could easily split this up into two parts: writing the parser and writing the cli. The cli could just pretend the parser already works and work with some artificial settings.

**Comment by akhandait** [comment]
@zoq What exactly will the parser do? Do you mean that the parser will actually implement all the necessary functions like forward and backward propagation and put them together in a model class and the CLI will just use this model like all the CLIs?
If so, then I would like to work on the parser as I have worked with ANNs and am quite comfortable with the nitty gritties.

**Comment by desai-aditya** [comment]
@akhandait , @zoq, @rcurtin - I think the parser will parse the file and return the values to the main CLI program. The values will then be used to make the model and train the dataset inside the CLI . There are some parameters that will be used in general that will be needed as arguments to the CLI. The parser should only parse the values and not compile the model. It could be that the user supplies the model too. This is what I understand. I have already started working on the CLI . @akhandait You can go ahead with the parser and join me as soon as you finish it.

**Comment by akhandait** [comment]
@desai-aditya Can you please have a look at the CLIs we already have. You can find them here:
src/mlpack/methods/<method_name>/<method_name_main.cpp>. 
All of them do the following job:
1) Info about the program - to be shown when somebody uses  `--help`.
2) Define all the parameters that the program will take from the terminal / python.
3) Throw appropriate errors/warnings for invalid parameters.
4) Pass the required parameters to a model of that method.
5) Run the model.
6) Return/store the required parameters/model.
Hope this helps. :)

**Comment by desai-aditya** [comment]
@akhandait  Don't worry I have already taken a look at how CLI's work and experimented with them. 
The only part where the trouble might start could be running and assembling the model since there's a  huge variety of different ways in which it can be built. Maybe you could help me with that once you've built the parser.

**Comment by akhandait** [comment]
Well, to the best of my knowledge and understanding, what I think a parser means here is that it will define the model and hence will have to take care of all the possible ways it can be built. The CLI, on the other hand, doesn't need to worry about the working of the model. As @zoq said, it just assumes that it works and uses it to get results.
@zoq Please correct me if I am wrong somewhere.

**Comment by zoq** [comment]
@desai-aditya is right, the parser will just parse the file that defines the structure and the CLI will build the model and run the model. So let's say we follow @rcurtin's idea of ``<layer type> <input size> <output size> <other parameters...>`` a simple example could be:

```
Linear 10 10
Sigmoid
Linear 10 5
Softmax
```

the cli builds the network based on the provided information:

```
FFN<NegativeLogLikelihood<> > model;
model.Add<Linear<> >(inputSizeA, outputSizeA);
model.Add<SigmoidLayer<> >();
model.Add<Linear<> >(inputSizeB, outputSizeB);
model.Add<LogSoftMax<> >();
```

and performs the requested action.

I hope this makes sense, let me know if I should clarify anything. A good starting point for the parser is: https://github.com/mlpack/mlpack/blob/master/src/mlpack/core/data/load_csv.hpp

**Comment by akhandait** [comment]
Okay, really sorry for the misinformation.
@desai-aditya it's good you have already started with the CLI program, I have my exams over the next week, I will still try to take as much time out as I can and work on the parser, I am also working on another issue, so I won't claim too many tasks as it will just delay necessary tasks.
When I am done with the parser and the other issue, I will be happy to help with the CLI program. :)

**Comment by zoq** [comment]
@akhandait don't worry, and best of luck with your exams.

**Comment by Namrata96** [comment]
Hi @desai-aditya @akhandait , are you still working on this? If yes, is there any way I could help?

**Comment by akhandait** [comment]
Hi @Namrata96. Yeah, I have been a little slow with regard to this issue but am working on it. I think I will open a WIP pull request in coming days and maybe you can help me by reviewing it extensively.

**Comment by desai-aditya** [comment]
@akhandait @zoq I am extremely sorry for delay in replying. I had my exams in the past week. I am free now and will be continuing back on the CLI. 
@akhandait Hows the parser working?
@Namrata96 As @akhandait said , you could help in reviewing the PR.

**Comment by Namrata96** [comment]
@desai-aditya @akhandait okay, sure.

**Comment by akhandait** [comment]
I saw on irc that a lot of people want to take this up. The proposal is taking a lot of my time recently and I will be okay if someone wants to work on the parser which I was working on. I will be happy to help in the reviewing of the PR. :)

**Comment by Rajiv2605** [comment]
@akhandait I would love to work on this! Please let me know if can work on something.

**Comment by cjlcarvalho** [comment]
Hi,

Is anyone working on this or anyone who had any progress?

Anyone have any thoughts about how the input should be read by the application?

Another question is about the output, do you think that only a JSON output for classification should be fine?

**Comment by cjlcarvalho** [comment]
@rcurtin: I am working on the parser. I have some questions about the CLI application, but I will try to work on it later after completing this parser problem, maybe I just need to take a deeper look at how the main code from the other CLI applications work. 

I'm thinking about a map between the layer type and the layer desired parameters to parse the information, similar to what was commented above.

**Comment by sreenikSS** [comment]
If this has not been completed yet, I would like to do it. According to what I have understood from the discussion above and related discussion on https://github.com/mlpack/mlpack/issues/1267 I am drawing out an implementation plan for myself and anyone else who might find it useful:

Format of the network configuration file: (say network.conf)

"loss-definition"
linear 50 30
relu
linear 20
relu
linear 5
softmax

This translates into:

FFN<NegativeLogLikelihood<> > model; // or the specified loss definition
model.Add<Linear<> >(50, 30);
model.Add<ReLULayer<> >();
model.Add<Linear<> >(30, 20);
model.Add<ReLULayer<> >();
model.Add<Linear<> >(20, 5);
model.Add<LogSoftMax<> >();

The complete list of layer types I suppose can be obtained from https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/layer/layer_types.hpp
(not too sure about this one, maybe the "layer" folder can be of better use, would want this reclarified)

The parser would use load_csv to load the file and call the appropriate methods to build the model.

The CLI would have two primary options:
1. Train the model given a dataset and the network.conf file (option to test/save the model and load weights from an existing model)
2. Test a dataset on a model(path to be provided)
(info regarding implemented CLIs for individual methods is available over here at https://github.com/mlpack/mlpack/issues/1254#issuecomment-367572549)

I am also working on the model translation with tensorflow, etc and the CLI could be later extended to run/convert those models too once that is done.
@zoq @rcurtin what are your thoughts regarding this? And please link any work that has already been done.

**Comment by rcurtin** [comment]
@sreenikSS what you've written sounds reasonable to me.  I don't know of any work that people have already done towards this, and I can't remember any PRs that were attempts.  Overall I think that the format you have is good; the key here will be documenting the format of `network.conf` extensively.

**Comment by sreenikSS** [comment]
Yes, the format needs to be decided upon before much progress can be made. I am digging into the total number of possibilities in order to maintain the entire functionality of the ANN module. I shall come up with a proposal shortly.

**Comment by sreenikSS** [comment]
### Allowed classes for the CLI

### Activations: (not obtained from layer)
identity_function
logistic_function
rectifier_function
softplus_function
softsign_function
swish_function
tanh_function
### Activations obtained from layer: (function prototypes given for more clarity)
```
ELU(const double alpha); // for elu

ELU(); // for selu

FlexibleReLU(const double alpha = 0);

HardTanH(const double maxValue = 1, const double minValue = -1);

LeakyReLU(const double alpha = 0.03);

PReLU(const double userAlpha = 0.03); // parametric relu
```

### Loss functions: (function prototypes)
```
CrossEntropyError(const double eps = 1e-10);

EarthMoverDistance();

KLDivergence(const bool takeMean = false);

MeanSquaredError();

NegativeLogLikelihood();

ReconstructionLoss();

SigmoidCrossEntropyError();
```


### Layers: (function prototypes)
```
AlphaDropout(const double ratio = 0.5,
               const double alphaDash = -alpha * lambda);

BatchNorm(const size_t size, const double eps = 1e-8);
BatchNorm();

Constant(const size_t outSize = 0, const double scalar = 0.0);

Convolution(const size_t inSize,
              const size_t outSize,
              const size_t kW,
              const size_t kH,
              const size_t dW = 1,
              const size_t dH = 1,
              const size_t padW = 0,
              const size_t padH = 0,
              const size_t inputWidth = 0,
              const size_t inputHeight = 0);

DropConnect(const size_t inSize,
              const size_t outSize,
              const double ratio = 0.5);

Dropout(const double ratio = 0.5);

FastLSTM(const size_t inSize,
           const size_t outSize,
           const size_t rho = std::numeric_limits<size_t>::max());

GRU(const size_t inSize,
      const size_t outSize,
      const size_t rho = std::numeric_limits<size_t>::max());

LayerNorm(const size_t size, const double eps = 1e-8);
LayerNorm();

LinearNoBias(const size_t inSize, const size_t outSize);

Linear(const size_t inSize, const size_t outSize);

 MaxPooling(const size_t kW,
             const size_t kH,
             const size_t dW = 1,
             const size_t dH = 1,
             const bool floor = true);

 MeanPooling(const size_t kW,
              const size_t kH,
              const size_t dW = 1,
              const size_t dH = 1,
              const bool floor = true);

MultiplyConstant(const double scalar = 1.0);

Recurrent(const StartModuleType& start,
            const InputModuleType& input,
            const FeedbackModuleType& feedback,
            const TransferModuleType& transfer,
            const size_t rho);

TransposedConvolution(const size_t inSize,
                        const size_t outSize,
                        const size_t kW,
                        const size_t kH,
                        const size_t dW = 1,
                        const size_t dH = 1,
                        const size_t padW = 0,
                        const size_t padH = 0,
                        const size_t inputWidth = 0,
                        const size_t inputHeight = 0);

```
These are the what I thought would make a good and practically usable CLI. I shall post the declaration definitions of each of the above functions for the `network.conf` file tomorrow along with some other clarifications I have.
Meanwhile I'll also devise a syntax for declaring a custom loss function using variables of the training dataset.
I know it is not easy to review all of the functions at once but please tell me if I have missed anything obvious.

**Comment by sreenikSS** [comment]
According to what I have thought, the ```network.conf``` file would be **case-insensitive** and there would be multiple *allowed declarations* for a single function. Eg: ```CrossEntropyError``` can be declared as ```crossentropy```, ```crossentropyloss```, ```cross_entropy_loss``` or ```crossentropyerror```.
### An approximate example of the ``` network.conf ``` file for generating something similar to the architecture in [this digit recogniser model](https://github.com/mlpack/models/blob/master/Kaggle/DigitRecognizer/src/DigitRecognizer.cpp):

loss NegativeLogLikelihood
init random -1 1
optimizer sgd adam_update 
step_size 0.0005
batch_size 50
train_test 0.1
linear 100
sigmoid
linear 100
sigmoid
linear 10
LogSoftMax

What I am confused with is the number of parameters in the optimizer, that is, whether to specify it in a single line like ```sgd adam_update 0.0005 50``` or in the broken up way I have shown here above (which does not contain all the parameters actually used by that program, I have left out information for simplicity). 
Moreover, I believe this cannot be planned out completely at the beginning, I think problems will crop up as the thing gets more and more developed and they would need to be solved therein.
Now that my exams are finally over, I shall get started right away. I might be having frequent doubts, should I post them here or to the IRC?

@rcurtin could you have a look here. One more thing, I have omitted some functions of ```layer``` because I wanted to keep the CLI simple at the beginning; examples of some  functions that have been omitted are ```atrous_convolution```, ```bilinear_interpolation``` among a few others. Should I add them?

**Comment by sreenikSS** [comment]
The main questions that arise (including repetitions from the previous comments) are:

- Will I implement **all** the functions that are present in the ann codebase? In the above comments I have mentioned most of the functions present in the codebase, though not all. Back at that time I was not very familiar with the codebase, but now I think I can implement all the functions present in the ann subsection.
- In ```network.conf```, the line ```input 1,0,0,-1``` (or just ```input``` for that matter) denotes the end of the parameter description and beginning of the model definition. Now when it comes to the parameter description part, shall I remove the flexibility of selecting custom hyperparameter values for the **update** function like ```AdamUpdate(...)``` and keep only the necessary ones like step size, batch size, etc. declarable?
- Do you think keeping multiple allowed declarations for a single component is practical enough? Eg: SGD/StochasticGradientDescent/stochastic_gradient_descent pointing to the same SGD component. The idea is to consider the **full form/short form** if both are popular, **underscores** between words and a completely **case-insensitive** approach.
- Allowed layer formats as a **json** or have them **hard-coded**? I have tried weighing the options, some of the pros and cons of having a **json configuration file** are:

PROS:
1. Can be easily understood and edited to accommodate more functions and update the existing ones.
2. (need to think about it)

CONS:
1. The json has to be distributed in the installation directory and might seem quite out of place.
2. (need to think about it)

Another thing that needs to be kept in mind is to choose the format that is more suitable for re-use in the tensorflow-->mlpack model translator.

tf to mlpack model generator: tf/pytorch/caffe model --> onnx model --> structured and loaded into the program variables --> mlpack model

CLI: CLI input linked with ```network.conf``` --> structured and loaded into the program variables --> mlpack model

An example of a subsection of the proposed json file (although the loss functions are pretty simple as they lack parameters):
```
{
    "loss": {
        
            "NegativeLogLikelihood": [
                "negative",
                "log",
                "likelihood",
                ["", "error", "loss"]
            ],
            "CrossEntropyError": [
                "cross",
                "entropy",
                ["", "error", "loss"]
            ],
            "EarthMoverDistance": [
                "earth",
                "mover",
                "distance"
            ],
            "KLDivergence": [
                ["k", "kullback"],
                ["l", "leibler"],
                "divergence"
            ]
        } 
    
}
```

**Comment by rcurtin** [comment]
Hey there @sreenikSS, sorry to make you wait two weeks for a response.  There is a lot going on and this is a time-consuming issue to answer because the design is an important step here, so it's hard to find a block of time to get into this to give a good answer. :)

I read through all the comments you've left and I like the design that you are settling on here, and that you've been thinking about the ONNX/TensorFlow translators so that people can import/export models.  That's an important thing that wasn't on my mind when I was originally thinking about this issue but it's definitely something that we should try to make easy if we can.

I also thought about this from another angle---bindings to languages other than the CLI.  Right now we can take in filenames, and that makes sense for the command-line; one would want to specify `network.conf`.  But from Python, it would make way more sense to just pass in, e.g., a `dict` that describes the structure of the network.  (Internally we can wrangle that into whatever format we need.)  This will also require supporting a `dict` type (or something like it) in other languages, but we can save that discussion for later---that support should not be a limiting factor in what we do here.

So maybe a JSON-like (or YAML-like) file structure is better.  Here's a proposed adaptation of the proposed digit recognizer network you gave (in JSON):

```
{
  "loss": "NegativeLogLikelihood",
  "init":
    {
      "type": "random",
      "min": "-1",
      "max": "1"
    }
  "optimizer":
    {
      "type": "adam",
      "step_size":  0.0005,
      "batch_size": 50
    }
  "network":
    [
      { "type": "linear", "units": 100 },
      { "type": "sigmoid" },
      { "type": "linear", "units": 100 },
      { "type": "sigmoid" },
      { "type": "linear", "units": 10 },
      { "type": "logsoftmax" }
    ]
}
```

Maybe that would look nicer in YAML?  Not sure.

Anyway, I don't know that we need to handle the train/test split or input format inside the network definition.  After all, the binding will have other inputs that can also be used to specify those things.

The list of layers you wanted to support seems totally reasonable to me.  It should be pretty easy to just add a layer to the binding by specifying a name to be used in the configuration, the name of the class, then the name (and type) of each constructor parameter for that layer (and whether it's required).

Let me know what you think or if I can clarify anything I've written here.  I hope that it's helpful. :)

**Comment by sreenikSS** [comment]
This sounds good. A JSON or YAML looks much more organised than a simple text file and suitable for future extension like you have mentioned. 

With JSON I was using boost's `json_parser` and `property_tree`, YAML support is currently in the pipeline but then there are other libraries that parse YAML, so it won't be a problem extending support to YAML regarding whatever we are doing with JSON. I prefer to start with JSON and maybe we can provide YAML support once a major block of work is finished.

Also, as we are making it much more organised, I feel there is no need for providing multiple allowed declarations for certain functions, like I was thinking about `cross_entropy/cross_entropy_loss/crossentropyloss`. According to the current plan, only say `CrossEntropy` will be allowed.

So I expect the user is required to be thorough with the docs, especially with the variable names like `"max"` and `"min"` for the `"type"` `random`.  That's fine because the documentation is neat enough for this purpose (though I am personally not a big fan of the current documentation we have, especially the tutorials).

This also solves the question regarding the maintenance of either a **JSON** or a **hard-coded** implementation for keeping track of the allowed keywords, declarations and parameters. I feel a JSON would be pretty much **out of place** and a **well-commented code is the way to go**.

@rcurtin just let me know if you are of the same opinion.

**Comment by rcurtin** [comment]
I thought about it a bit more; it may be useful to think about using YAML instead of JSON first since it's a lot easier to handwrite YAML.  But I am not too picky---both work, and we can also write other tools on top of it all that produce JSON/YAML from an "easier" representation.

> According to the current plan, only say CrossEntropyError will be allowed (case-insensitive though).

Agreed, that's fine with me.

> So I expect the user is required to be thorough with the docs, especially with the variable names like "max" and "min" for the "type" random. That's fine because the documentation is neat enough for this purpose (though I am personally not a big fan of the current documentation we have, especially the tutorials).

Do you think the new http://www.mlpack.org/doc/mlpack-3.0.4/cli_documentation.html might be a better thing to do here?  We could embed a full description of the language and format either in the `Detailed documentation` section of the binding, like this:

http://mlpack.org/doc/mlpack-3.0.4/cli_documentation.html#approx_kfn_detailed-documentation

Or have a couple excerpts of "common" usage, and then refer to a tutorial that's written somewhere else.  What do you think?

> This also solves the question regarding the maintenance of either a JSON or a hard-coded implementation for keeping track of the allowed keywords, declarations and parameters. I feel a JSON would be pretty much out of place and a well-commented code is the way to go.

I'd agree, I feel like we can just have some C++ code that, perhaps, extracts the allowed layer types from `LayerTypes` in `src/mlpack/methods/ann/layer/layer_types.hpp`.  This might be worth thinking about a little bit more: given the template type `LayerTypes`, can we extract a list of the members, and then also extract related documentation for each of the parameters to that layer?  Or will we have to maintain that separately?  I don't like maintaining more than what we need to, but there are probably limitations on what we could do with crazy template metaprogramming techniques to parse the `LayerTypes` type.  Let me know what you think... I don't have a well-defined opinion on what the best way to handle the documentation here is.

**Comment by sreenikSS** [comment]
> I thought about it a bit more; it may be useful to think about using YAML instead of JSON first since it's a lot easier to handwrite YAML. But I am not too picky---both work, and we can also write other tools on top of it all that produce JSON/YAML from an "easier" representation.

@rcurtin you are right, YAML is more user-friendly but I have just finished the JSON implementation with a boost property tree storing it and categorising the information into a number of maps storing the data in a dictionary like format. It am all for YAML but I currently want to focus on making the whole thing work; I shall provide YAML support (not much code though) after the main part is done. What do you say?

> Do you think the new http://www.mlpack.org/doc/mlpack-3.0.4/cli_documentation.html might be a better thing to do here? We could embed a full description of the language and format either in the Detailed documentation section of the binding, like this:
> 
> http://mlpack.org/doc/mlpack-3.0.4/cli_documentation.html#approx_kfn_detailed-documentation
> 
> Or have a couple excerpts of "common" usage, and then refer to a tutorial that's written somewhere else. What do you think?

I have tried following the format of the original documentation, but there are some exceptions (for better readability), for example the linear layer has the original documentation:
`Linear(const size_t inSize, const size_t outSize)`
But we have abstracted the `inSize`, so the user only needs to specify `outSize`, which is better expressed as:
`"type": "linear", "units": 100` 
because "units" sounds more appropriate than "outSize". But these instances are less in number, so maintaining a separate set of documentation is probably not needed. A separate tutorial mentioning those instances should be sufficient.

> I'd agree, I feel like we can just have some C++ code that, perhaps, extracts the allowed layer types from LayerTypes in src/mlpack/methods/ann/layer/layer_types.hpp. This might be worth thinking about a little bit more: given the template type LayerTypes, can we extract a list of the members, and then also extract related documentation for each of the parameters to that layer? Or will we have to maintain that separately?

You'll be glad to hear, that part is already done and I have done it exactly as you have mentioned here. I am structuring it in such a way that bindings and other stuff would be much easier to add. Actually I am learning a lot (A lot actually!!) while building this CLI.

One more thing, regarding the documentation are we planning to participate in Google's newly announced Season of Docs this year?

**Comment by rcurtin** [comment]
> It am all for YAML but I currently want to focus on making the whole thing work; I shall provide YAML support (not much code though) after the main part is done. What do you say?

That's fine with me. :+1:

> I have tried following the format of the original documentation, but there are some exceptions (for better readability), for example the linear layer has the original documentation:
> `Linear(const size_t inSize, const size_t outSize)`
> But we have abstracted the inSize, so the user only needs to specify outSize, which is better expressed as:
> `"type": "linear", "units": 100`
> because "units" sounds more appropriate than "outSize". But these instances are less in number, so maintaining a separate set of documentation is probably not needed. A separate tutorial mentioning those instances should be sufficient.

Ah, what I was suggesting here is automatically generating the documentation for each layer, as opposed to manually writing it.  It may be complex to come up with a good solution for that.

> You'll be glad to hear, that part is already done (the code, not the doc part though) and I have done it exactly as you have mentioned here. I am structuring it in such a way that bindings and other stuff would be much easier to add. Actually I am learning a lot (A lot actually!!) while building this CLI.

Great!  If there is any part of the code I can explain, just let me know.  There are a lot of complex pieces that fit together in complex ways. :)

> One more thing, regarding the documentation are we planning to participate in Google's newly announced Season of Docs this year?

Not sure, I think there is some interest in it in the community, so we'll see. :+1:  I think it is a cool program, the issue for me at least is always time.  (That said I don't have to run the effort, so if there's critical mass to do it, I think we should!)  :)

**Comment by sreenikSS** [comment]
Thanks. I shall let you know if I need any further clarification anywhere.

**Comment by sreenikSS** [comment]
@rcurtin the parser is almost done. There are a number of critical parameters that need to be addressed. I shall post them tomorrow (extremely tired right now :( ) 

The PR is over here:
https://github.com/mlpack/mlpack/pull/1837

**Comment by sreenikSS** [comment]
Critical comments on it:
1. The init type doesn't take the given parameters into account because it is passed as a template parameter but not as an argument. The fix is quite simple. I am a little tight on schedule but will fix it as soon as I find some time.
2. For model training and testing the code needs to be modified every time one switches from a classification problem to a regression problem. This can be solved by either taking a user input (when the CLI is complete) or by determining it from the last layer and given labels (i.e., the matrix 'trainY').
3. Not all features have been implemented yet. Additionally, some init types and layer types have been commented because my mlpack installation somehow ignored them so it won't find the required files and hence I cannot compile and test it.
4. Programs in general have multiple methods and they are called in sequence from a separate method (main() in most standalone applications). But here due to the absence of suitable return types, the functions call one another in a chain like fashion, i.e., say main() calls A(), A() calls B(), B() calls C() and so on. Maybe an inheritance based solution can help by having a superclass to all init types, optimizer types and loss types respectively just like LayerTypes is a superclass to all layers (does it already exist?)
5. The functions are named like getters, like getLossType() and getInitType() whereas they are simple void functions and are even part of the chain discussed in point 4. These names are given to keep the essence of these methods in mind and the ideal behaviour they could exhibit (if a suitable return type is available).
6. The style checks do not currently pass, I hope to update the code soon.

I will update this if something else comes to my mind.

**Comment by kunakl07** [comment]
I would like to work on this issue. Can I take this issue?

**Comment by bkmgit** [comment]
This would be a useful addition. A CLI is already part of [nnabla](https://github.com/sony/nnabla/blob/master/doc/python/command_line_interface.rst) 

Is anybody still working on this?

**Comment by RishabhGarg108** [comment]
Hey, I was recently looking at this issue. After reading the discussion here, I figured out that we need to do two things here - 
1. A utility to parse the config file to build the model.
2. An actual CLI that would implement that model.

In my opinion, a simple text file would not suffice, because it is easier to think of the model in a structured way (json/xml)

Then, key question is "How do we parse that structured config file?"

@sreenikSS has almost completed the parser in #1837 , but it is done using `boost::property_tree`. Since, we are trying to reduce our dependence on boost libraries, I thought we need to do it in some other way. I discussed it on IRC with @zoq . He suggested me to use cereal for it. So, I gave it a shot and tried to see if that works out or not, because cereal can be used to save or load objects into different formats like json/xml. 

So, to test it out, I created a small model and saved it in the following way.
FFN<> model;
model.Add<Linear<> >(20, 3);
model.Add<SigmoidLayer<> >();
model.Add<LogSoftMax<> >();

data::Save("model.json", "model", model, false);

The model.json generated looks like [this](https://drive.google.com/file/d/1KZxO1jX8aHyt1jgKPzkIH7nOH0_7bGme/view?usp=sharing). This looks pretty bad because this actually saves all the parameters and everything associated to the model. To use cereal, the user actually needs to make a config file similar to it, This is practically not possible for user to write it.

So, now we are back to the same question i.e. how do we parse config file.
We can do it two ways - 
1. By writing our own json/xml parser. (which in my opinion would be a tedious task)
2. By using some external dependency. (which would increase external dependence)

So, I want to ask what should be a good way to approach this? Thanks.

**Comment by zoq** [comment]
What I had in my mind was to have some sort of config class that we could serialize and use it as an intermediate format to construct the model. Isn't that what the parse would do as well? I would expect that a config class something like:

```c++
class Config
{
 public:
   std::vector<struct{std::string name, inSize, outSize}> layerName;
}
```

Would look easy enough.

**Comment by bkmgit** [comment]
@RishabhGarg108 The following may be relevant:
https://www.khronos.org/nnef/
https://github.com/onnx/onnx/blob/master/docs/IR.md

**Comment by RishabhGarg108** [comment]
@bkmgit , thanks for these resources. I will definitely look at them.
The ONNX thing looks great, because that way we can import models from all those other libraries that ONNX support.

I have no experience with parsing and stuff. So currently, I am trying and exploring various things, like what @zoq mentioned about a `Config` class etc.. I will also give ONNX a try and we will see what works best. :+1:

**Comment by RishabhGarg108** [comment]
> @RishabhGarg108 The following may be relevant:
> https://www.khronos.org/nnef/
> https://github.com/onnx/onnx/blob/master/docs/IR.md

@bkmgit , I looked at both of these. It turns out that both of these tools are for interoperability and optimization of trained models across different frameworks and devices for making inferences. This is not exactly what we are looking for. Infact, all we want is just a simple way to define the architecture for an ann model and then be able to translate that into a mlpack ann model. I hope that makes sense.
Correct me if I overlooked something :D

**Comment by NippunSharma** [comment]
@rcurtin @zoq Can I work on this issue, it has been inactive for some time now ?

**Comment by rcurtin** [comment]
We should leave this open---it is still important functionality we should add at some point.

**Comment by ranjan2829** [comment]
A CLI (Command Line Interface) for ANN (Artificial Neural Networks) would be a useful tool for developers and researchers who want to easily train and test ANN models using the command line. Here are some of the features that a CLI for ANN could include:

Training and testing models: The CLI should allow users to train and test ANN models using different datasets and configurations. Users should be able to specify the type of model they want to train, the dataset they want to use, and the hyperparameters they want to tune.

Saving and loading models: The CLI should allow users to save and load trained models to disk, so they can reuse the models later for prediction or further training.

Visualizing training progress: The CLI should provide visual feedback on the progress of training, such as plots of loss and accuracy metrics over time.

Support for different ANN architectures: The CLI should support different ANN architectures, such as feedforward neural networks, convolutional neural networks, and recurrent neural networks.

Integration with other libraries: The CLI should integrate with other libraries that are commonly used for ANN, such as TensorFlow, PyTorch, and Keras.

Command line arguments and options: The CLI should allow users to specify command line arguments and options to customize the training process. For example, users could specify the learning rate, the number of epochs, or the batch size.

Cross-platform compatibility: The CLI should be compatible with different operating systems, such as Linux, macOS, and Windows.

Some of the challenges that you may face while building a CLI for ANN include:

Ensuring that the CLI is easy to use: The CLI should have a clear and simple interface that is easy to use, even for users who are not familiar with ANN.

Handling different types of data: ANN models can be trained on different types of data, such as images, audio, or text. The CLI should be able to handle different types of data and preprocess the data appropriately.

Integrating with different ANN libraries: The CLI should integrate with different ANN libraries, which can be a challenging task since each library has its own API and interface.

Debugging and testing: It can be challenging to debug and test a CLI, especially when dealing with large datasets and complex models.

Overall, building a CLI for ANN can be a challenging but rewarding task, as it can greatly improve the accessibility and usability of ANN models for developers and researchers.

**Comment by kakarrot95** [comment]
Ok so , i know C++ , and I am studying ML , and i think i might be able to work on this , but maybe anyone just give me a little guide ? .

**Comment by SANJITH-KUMAR-20** [comment]
is this issue still open?

**Comment by Shudhanshu25** [comment]
If you want to train or use an ANN from the CLI, you typically use scripts or frameworks that expose CLI functionality. Here’s how you can do it:

Using Predefined Scripts
Write a Python script to define, train, and save your ANN model.
Run it from the CLI with arguments, e.g.:
python train_ann.py --input_dim 100 --output_dim 10 --epochs 20

**Comment by Blackcipher101** [comment]
is this issue still open to work on?

**Comment by prtksh** [comment]
Hello, I would like to contribute to this feature. May I know what's the update on this till now? Or maybe even some more implementation details discussed before. Thanks.

**Comment by coderuhaan2004** [comment]
Hi,

I’ve gone through the discussion above and would like to continue working on this issue. I'm planning to structure the implementation using a namespace that includes `enum class` for different types of initialization rules, loss functions, and layer types, along with structs like `initConfig`, `lossConfig`, etc., to store the selected types and their parameters. I think this approach adds modularity and clarity to the code.

The JSON config will be parsed using `Boost.PropertyTree`, and the extracted values will be mapped to structs which are passed to the main function and will be used to initialize components like the `FFN`, optimizer, etc.

Would like to get insights on it whether this is aligned with the intent of the functionality.

Thanks!

**Comment by ranjodhsingh1729** [comment]
Hi everyone,

I went through the previous discussions and related PRs. From what I can tell, it doesn’t seem like anyone is currently working on it—though if someone is, please do let me know!

I understand this has been a long-standing and tricky feature to get right. I’ve read the earlier ideas and really appreciate the thought that’s gone into it—especially the five points mentioned by @rcurtin, which I think are non-negotiable.

Here are a few thoughts based on what I’ve seen so far:

1. **Config Format:** JSON seems like a good starting point because it’s simple and well-supported. A limited subset of YAML might offer a better user experience later, but writing a clean and reliable parser could take time. Starting with JSON and adding YAML support later seems like a practical approach.

2. **Parameter Precedence:** The config should let users define the full network and its hyperparameters, while also allowing command-line overrides. The priority should be: `CLI > config file > defaults`.

3. **Data Handling Options:** It would be helpful to include options for data loading—like file paths, batch sizes, and preprocessing—in the config, with CLI overrides where needed.

4. **Advanced YAML Features (Optional Later):** Features like anchors or aliases could be useful for complex models, but can be added later once the basics are in place.

5. **Documentation:** This feature would need solid documentation with examples so that users can easily understand and use it.

While going through the past work, I came across [PR #1837](https://github.com/mlpack/mlpack/pull/1837) by @sreenikSS. It includes some solid work using JSON that could serve as a good starting point. I saw that it was closed with a note mentioning that—based on a discussion I couldn't find—it was decided this should be handled in a separate model converter repo. From what I understand, the goal is to make the CLI interoperable so that models can be saved and loaded using import/export formats compatible with other libraries most probably using ONNX. I assume that this is in addition to our own configuration format (YAML OR JSON).

I’m interested in picking this up, but I want to be sure there's still interest in adding this feature to the main repo before putting in too much work. Happy to get feedback.

Thanks!

**Comment by greendaffodils** [comment]
What is the update and is our current final goal.And what all do we want to implement.
Some highlighted point to keep in mind while building this parser like being fast not depending on libraries anything else.I want to work on this .

**Comment by rcurtin** [comment]
As much as I would like this support to get done, it needs to be done the right way and there currently isn't bandwidth to do it correctly, so let's close it.  We can return to it someday, eventually.
